### DISCUSSION TITLE
Substitute the Brainnet CNN by Pyg

### DISCUSSION THREAD

**Discussion by yikang7720** [pyg-team/pytorch_geometric#4609] [Q&A]
Hi,

I am now reading the paper about the Brainnet CNN. The Brainnet CNN is an CNN archiecture which performs extremly well for brain graph tasks. However, the Brainnet CNN was implemented in 2017. Then I am wondering if there is any way that we can substitute the  Brainnet CNN by GNN. The brainnet CNN consists of three main layers, edge to edge convolution, edge to node convolution and node to graph convolution. For example, the main idea edge to edge convolution is similar to a 3 × 3 × M` convolution filter over a stack of 2D grid data, in that, for each feature map, it combines the signal at some point with the signal from the direct neighbours. I am also assuming that edge to edge convolution is more likely to make convs to the edge features of the adjacency matrix. 

Here is the code of edge to edge convolution. 

`
Class E2EBlock(torch.nn.Module):

    def __init__(self, in_planes, planes,bias=True):
        super(E2EBlock, self).__init__()
        self.d = 116 #Number of nodes
        self.cnn1 = torch.nn.Conv2d(in_planes,planes,(1,self.d),bias=bias)
        self.cnn2 = torch.nn.Conv2d(in_planes,planes,(self.d,1),bias=bias)       
    def forward(self, x):
        #Input is adjacency matrix
        a = self.cnn1(x)
        b = self.cnn2(x)
        return torch.cat([a]*self.d,3)+torch.cat([b]*self.d,2)`

Here is the code of Brainnet CNN.

`Class Brainnet(torch.nn.Module):

    def __init__(self):
        super(Discriminator, self).__init__()
        self.d = 90
        self.e2econv1 = E2EBlock(1,32)   ######### Changed input channel from 1 to 3
        self.e2econv2 = E2EBlock(32,64)
        self.E2N = torch.nn.Conv2d(64,1,(1,self.d))   #Edge to node convolution
        self.N2G = torch.nn.Conv2d(1,256,(self.d,1))   #Node to graph convolution
        self.dense1 = torch.nn.Linear(256,128)
        self.dense2 = torch.nn.Linear(128,30)
        self.dis = torch.nn.Linear(30,1)
      
    def forward(self, x):
        out = F.leaky_relu(self.e2econv1(x),negative_slope=0.33)
        out = F.leaky_relu(self.e2econv2(out),negative_slope=0.33) 
        out = F.leaky_relu(self.E2N(out),negative_slope=0.33)
        out = F.leaky_relu(self.N2G(out),negative_slope=0.33)
        out = out.view(out.size(0), -1)
        out = F.leaky_relu(self.dense1(out),negative_slope=0.33)
        out = F.leaky_relu(self.dense2(out),negative_slope=0.33)
        out = self.dis(out)  
        return out`

Here is the architecture of Brainnet CNN.
![微信图片_20220509225808](https://user-images.githubusercontent.com/92366744/167505810-c10c2508-8a3c-43c5-b827-d804390a6b8f.png)

Here is the link to the Brainnet CNN.
[https://www2.cs.sfu.ca/~hamarneh/ecopy/neuroimage2017.pdf](url)

**Comment by xuesongwang** [comment]
Hi there,
 Yes indeed some researchers have been doing these for a while, please check this repo: https://github.com/LifangHe/BrainGNN_Pytorch
  and this review paper: https://arxiv.org/pdf/2106.03535.pdf  (Graph Neural Networks in Network Neuroscience)
 Hope that's hepful.
